pkg/runtime.v2/framework/plugins/plainml/plainml.go

```diff
@@ -40,16 +41,34 @@ func (p *PlainML) Name() string {
 	return Name
 }
 
-func (p *PlainML) EnforceMLPolicy(info *runtime.Info) error {
+func (p *PlainML) EnforceMLPolicy(info *runtime.Info, trainJob *kubeflowv2.TrainJob) error {
 	if info == nil || info.MLPolicy == nil || info.MLPolicy.Torch != nil || info.MLPolicy.MPI != nil {
 		return nil
 	}
-	numNodes := ptr.Deref(info.MLPolicy.NumNodes, 1)
+
+	// TrainJob contains the actual information for the number of nodes.
+	numNodes := info.MLPolicy.NumNodes
+	if trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.NumNodes != nil {
+		numNodes = trainJob.Spec.Trainer.NumNodes
+	}
+	info.Trainer.NumNodes = numNodes
+
+	// Add envs from the TrainJob.
+	if trainJob.Spec.Trainer != nil {
+		info.Trainer.Env = append(info.Trainer.Env, trainJob.Spec.Trainer.Env...)
+	}
+
+	// Update total Pod requests for the PodGroupPolicy plugin.
 	for rName := range info.TotalRequests {
-		info.TotalRequests[rName] = runtime.TotalResourceRequest{
-			Replicas:    numNodes,
-			PodRequests: info.TotalRequests[rName].PodRequests,
+		// For other Jobs like the Initializer, replica is always equal to 1.
+		// TODO (andreyvelich): Add support for total requests from the TrainJob's ResourcesPerNode.
+		if rName == constants.JobTrainerNode {
```

Comment:
What is the reason that we store the totalRequests only for Trainer?

Reply:
Actually, we store totalRequest for other ReplicatedJobs as well.
The replicas are always 1 for now: 
https://github.com/kubeflow/training-operator/blob/f00c1d37ef03d89214ddd9a7189f98f81fce2276/pkg/runtime.v2/core/trainingruntime.go#L106-L109

Reply:
I meant that the reason why we do not store thre totalRequests for the initializers?

Reply:
We store totalRequests for the Initializer only from the TrainingRuntime template here: https://github.com/kubeflow/training-operator/blob/f00c1d37ef03d89214ddd9a7189f98f81fce2276/pkg/runtime.v2/core/trainingruntime.go#L106-L109

We create another entity in the PodSpecReplicas
https://github.com/kubeflow/training-operator/blob/f00c1d37ef03d89214ddd9a7189f98f81fce2276/pkg/runtime.v2/runtime.go#L73-L77

TrainJob can't override resources for the Dataset or Model initializers. 

Does it make sense @tenzen-y ?

Reply:
> TrainJob can't override resources for the Dataset or Model initializers.

I was not sure the reason why the Dataset initializer can not override those since we can pass the replica name like trainer, initializer and so on in the WithPodSpecReplicas, then we can modify it in each framework plugin.

Reply:
It's a good point, @tenzen-y however we don't support override for resources in PodSpecOverride: https://github.com/kubeflow/training-operator/blob/master/pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go#L243.
We can discuss it in the future, if we need to support resources in the override as well.

Reply:
Alright. Let's see this later. I guess that @shravan-achar can take this.